File: budgetease_flutter/lib/services/income_analyzer.dart
import 'dart:math';
import '../models/transaction.dart';
import '../models/income_pattern.dart';
import 'database_service.dart';

class IncomeAnalyzer {
  /// Analyse le pattern de revenus sur les 14 derniers jours
  static IncomePattern analyzeIncomePattern() {
    final now = DateTime.now();
    final twoWeeksAgo = now.subtract(const Duration(days: 14));

    final incomes = DatabaseService.transactions.values
        .where((t) => t.type == 'income' && t.date.isAfter(twoWeeksAgo))
        .toList();

    if (incomes.isEmpty) {
      // Pas de revenus récents : mode prudent
      return IncomePattern(
        estimatedWeeklyIncome: 0,
        minimumObserved: 0,
        averageObserved: 0,
        observationDays: 14,
        lastUpdated: now,
        isRegular: false,
      );
    }

    final total = incomes.fold(0.0, (sum, t) => sum + t.amount);
    final average = total / 2; // Moyenne sur 2 semaines
    final minimum = incomes.map((t) => t.amount).reduce(min);

    // Détection régularité : écart-type faible
    final variance = _calculateVariance(incomes.map((t) => t.amount).toList());
    final isRegular = variance < (average * 0.3); // Écart < 30%

    // Estimation prudente : utiliser le minimum si irrégulier
    final estimate = isRegular ? average : minimum;

    return IncomePattern(
      estimatedWeeklyIncome: estimate,
      minimumObserved: minimum,
      averageObserved: average,
      observationDays: 14,
      lastUpdated: now,
      isRegular: isRegular,
    );
  }

  /// Récupère ou crée le pattern stocké
  static IncomePattern getOrCreatePattern() {
    final box = DatabaseService.incomePatterns;
    
    if (box.isEmpty) {
      final pattern = analyzeIncomePattern();
      box.add(pattern);
      return pattern;
    }

    final pattern = box.values.first;
    
    // Mettre à jour si plus de 24h
    if (DateTime.now().difference(pattern.lastUpdated).inHours > 24) {
      final updated = analyzeIncomePattern();
      pattern.estimatedWeeklyIncome = updated.estimatedWeeklyIncome;
      pattern.minimumObserved = updated.minimumObserved;
      pattern.averageObserved = updated.averageObserved;
      pattern.observationDays = updated.observationDays;
      pattern.lastUpdated = updated.lastUpdated;
      pattern.isRegular = updated.isRegular;
      pattern.save();
    }

    return pattern;
  }

  /// Calcule la variance d'une liste de montants
  static double _calculateVariance(List<double> amounts) {
    if (amounts.isEmpty) return 0;
    
    final mean = amounts.reduce((a, b) => a + b) / amounts.length;
    final squaredDiffs = amounts.map((x) => pow(x - mean, 2));
    
    return squaredDiffs.reduce((a, b) => a + b) / amounts.length;
  }

  /// Estime le prochain revenu probable et sa date
  static Map<String, dynamic> estimateNextIncome() {
    final pattern = getOrCreatePattern();
    
    if (pattern.estimatedWeeklyIncome == 0) {
      return {
        'amount': 0.0,
        'daysUntil': 7,
        'confidence': 'low',
      };
    }

    // Si régulier : prochaine semaine
    // Si irrégulier : estimation prudente sur 2 semaines
    final daysUntil = pattern.isRegular ? 7 : 14;
    
    return {
      'amount': pattern.estimatedWeeklyIncome,
      'daysUntil': daysUntil,
      'confidence': pattern.isRegular ? 'high' : 'medium',
    };
  }
}

File: src/lib/income-analyzer.ts
import { db } from './db';
import { Transaction, IncomePattern } from '../types';

const PATTERN_ID = 'current';

/**
 * Calculate variance of an array of numbers
 */
function calculateVariance(numbers: number[]): number {
    if (numbers.length === 0) return 0;
    const mean = numbers.reduce((a, b) => a + b, 0) / numbers.length;
    const squaredDiffs = numbers.map(x => Math.pow(x - mean, 2));
    return squaredDiffs.reduce((a, b) => a + b, 0) / numbers.length;
}

/**
 * Analyze income pattern over last 14 days
 */
export async function analyzeIncomePattern(): Promise<IncomePattern> {
    const now = new Date();
    const twoWeeksAgo = new Date(now);
    twoWeeksAgo.setDate(twoWeeksAgo.getDate() - 14);

    const incomes = await db.transactions
        .where('date')
        .above(twoWeeksAgo)
        .and(t => t.type === 'income')
        .toArray();

    if (incomes.length === 0) {
        // No recent income: prudent mode
        return {
            id: PATTERN_ID,
            estimatedWeeklyIncome: 0,
            minimumObserved: 0,
            averageObserved: 0,
            observationDays: 14,
            lastUpdated: now,
            isRegular: false,
        };
    }

    const total = incomes.reduce((sum, t) => sum + t.amount, 0);
    const average = total / 2; // Average over 2 weeks
    const minimum = Math.min(...incomes.map(t => t.amount));

    // Detect regularity: low variance
    const amounts = incomes.map(t => t.amount);
    const variance = calculateVariance(amounts);
    const isRegular = variance < (average * 0.3); // Variance < 30%

    // Prudent estimate: use minimum if irregular
    const estimate = isRegular ? average : minimum;

    return {
        id: PATTERN_ID,
        estimatedWeeklyIncome: estimate,
        minimumObserved: minimum,
        averageObserved: average,
        observationDays: 14,
        lastUpdated: now,
        isRegular,
    };
}

/**
 * Get or create income pattern
 */
export async function getOrCreateIncomePattern(): Promise<IncomePattern> {
    const existing = await db.income_patterns.get(PATTERN_ID);

    if (existing) {
        // Update if more than 24h old
        const hoursSinceUpdate = (Date.now() - new Date(existing.lastUpdated).getTime()) / (1000 * 60 * 60);
        if (hoursSinceUpdate > 24) {
            const updated = await analyzeIncomePattern();
            await db.income_patterns.put(updated);
            return updated;
        }
        return existing;
    }

    // Create new pattern
    const pattern = await analyzeIncomePattern();
    await db.income_patterns.add(pattern);
    return pattern;
}

/**
 * Get estimated monthly income
 */
export function getEstimatedMonthlyIncome(pattern: IncomePattern): number {
    return pattern.estimatedWeeklyIncome * 4.33;
}

/**
 * Get confidence level (0-1)
 */
export function getConfidenceLevel(pattern: IncomePattern): number {
    if (pattern.observationDays < 7) return 0.3;
    if (pattern.observationDays < 14) return 0.6;
    if (pattern.isRegular) return 0.9;
    return 0.7;
}

/**
 * Estimate next income
 */
export function estimateNextIncome(pattern: IncomePattern): {
    amount: number;
    daysUntil: number;
    confidence: 'low' | 'medium' | 'high';
} {
    if (pattern.estimatedWeeklyIncome === 0) {
        return {
            amount: 0,
            daysUntil: 7,
            confidence: 'low',
        };
    }

    const daysUntil = pattern.isRegular ? 7 : 14;

    return {
        amount: pattern.estimatedWeeklyIncome,
        daysUntil,
        confidence: pattern.isRegular ? 'high' : 'medium',
    };
}
